### PR TITLE
Add Titan internal stats and metrics for blob file and live value

### PR DESCRIPTION
**Pull request by tabokie**
Titan InternalStats provides a global counter that can decrement as well as increment. Thread safety is guaranteed by plain atomic variable.

Add three metrics with the use of InternalStats:

* LIVE_BLOB_SIZE: total size of value referenced from LSM tree.
* NUM_BLOB_FILE: number of blob file.
* BLOB_FILE_SIZE: total size of blob file.

Also, to save parameter passing, introduce TitanStats as a wrapper around RocksDB Statistics and Titan's internal stats.

### CODE REVIEW

File: src/titan_stats.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    if (p == internal_stats_.end()) {
+      return nullptr;
+    } else {
+      return p->second.get();
+    }
+  }
+  TitanInternalStats* internal_stats() { return internal_stats(default_cf_); }
+
+ private:
+  Statistics* stats_ = nullptr;
```
Comment (yiwu-arbug) [inline src/titan_stats.h]: Let's not combine `Statistics` with `TitanStats`. From TiKV for both plain RocksDB and Titan, we pass `Statistics` through `DBOptions`. With the change here, we need to tell apart whether we are using plain RocksDB and Titan, and pass `Statistics` object differently, which make the interface more cumbersome.

Comment (tabokie) [reply src/titan_stats.h]: Remove TitanStats from options to clear the trouble.

File: src/titan_stats.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  Statistics* statistics() { return stats_; }
+  TitanInternalStats* internal_stats(uint32_t cf_id) {
+    auto p = internal_stats_.find(cf_id);
+    if (p == internal_stats_.end()) {
+      return nullptr;
+    } else {
+      return p->second.get();
+    }
+  }
+  TitanInternalStats* internal_stats() { return internal_stats(default_cf_); }
```
[the remark this answers is not in the document]
Comment (tabokie) [reply src/titan_stats.h]: It does support. This is a helper getter for those modules without knowledge of cf_id.

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/titan_stats.h]: All properties we added so far are CF specific (e.g. "num-blob-file" should return number of blob files for a specific CF, and so on so forth). `internal_stats()` and all overload of `AddStats`, `SubStats` and `ResetStats` without CF id can be removed.

File: src/blob_storage.cc [hunk 3]
```diff
@@ -45,6 +45,11 @@ void BlobStorage::ExportBlobFiles(
 void BlobStorage::AddBlobFile(std::shared_ptr<BlobFileMeta>& file) {
   WriteLock wl(&mutex_);
   files_.emplace(std::make_pair(file->file_number(), file));
+  AddStats(stats_, cf_id_, TitanInternalStats::LIVE_BLOB_SIZE,
```
Comment (yiwu-arbug) [inline src/blob_storage.cc]: Will all these properties be reconstructed on DB reopen?

File: src/db_impl.cc [hunk 4]
```diff
@@ -603,18 +631,23 @@ void TitanDBImpl::OnCompactionCompleted(
...
+      delta += -bfs.second;
       auto file = bs->FindFile(bfs.first).lock();
       if (!file) {
         // file has been gc out
         continue;
       }
       file->AddDiscardableSize(static_cast<uint64_t>(-bfs.second));
     }
+    SubStats(stats_.get(), compaction_job_info.cf_id,
+             TitanInternalStats::LIVE_BLOB_SIZE, delta);
```
Comment (yiwu-arbug) [inline src/db_impl.cc]: should only update stat if the file is not obsolete

File: src/db_impl.cc [hunk 5]
```diff
@@ -603,18 +631,23 @@ void TitanDBImpl::OnCompactionCompleted(
...
     }
 
+    uint64_t delta = 0;
     for (const auto& bfs : blob_files_size) {
+      delta += bfs.second;
       // blob file size < 0 means discardable size > 0
       if (bfs.second >= 0) {
         continue;
       }
+      delta += -bfs.second;
```
Comment (yiwu-arbug) [inline src/db_impl.cc]: The logic around `delta` doesn't look right to me. Line 636 should be remove I think.

Comment (tabokie) [reply src/db_impl.cc]: (how careless am i)

File: src/titan_stats.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    return false;
+  }
+
+ private:
+  static const std::unordered_map<std::string, TitanInternalStats::StatsType>
+      stats_type_string_map;
+  std::atomic<uint64_t> stats_[INTERNAL_STATS_ENUM_MAX];
+};
+
+class TitanStats {
```
[the remark this answers is not in the document]
Comment (tabokie) [reply src/titan_stats.h]: Seperating them is ok. I just don't want to make room for ugly interface, like passing a rocksdb::stats to A while passing a titan::stats to B.

File: src/blob_storage.cc (hunk 3 above)
Comment (tabokie) [reply src/blob_storage.cc]: I think they are. During `Apply` in `VersionSet::Recover`.

File: src/titan_stats.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (tabokie) [reply src/titan_stats.h]: If `statistics(stats_)` is too tedious, we can patch `RecordTick` instead. Your call.

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/titan_stats.h]: ok. let's replace `RecordTick` with another helper method to do `RecordTick(stats_->statistics(),...)`.

File: src/blob_storage.cc (hunk 3 above)
Comment (yiwu-arbug) [reply src/blob_storage.cc]: Hmm, it seems to me `VersionSet::Apply` doesn't reset `discardable_size`, so `live-blob-size` will be incorrect anyway. Let's address the issue separately.

File: src/blob_storage.cc [hunk 7]
```diff
@@ -45,6 +45,11 @@ void BlobStorage::ExportBlobFiles(
...
+           file->file_size());
```
Comment (Connor1996) [inline src/blob_storage.cc]: seems it doesn't follow the definition `total blob value size referenced by LSM tree`. `file_size` includes header + footer + key

**Comment by tabokie** [comment]
Latest fix should resolve build failure on master branch, PTAL @yiwu-arbug

**Pull request merged**